### PR TITLE
Adds TCP info on Windows

### LINKED ISSUE #12305: Per-request rtt latency metrics in HTTP filters
The use-case I’d like to address is the ability to measure the rtt latency to upstreams so that this per-request rtt latency can be reported in detailed metrics.

Within the Linux TCP/IP stack, the TCP-level rtt is computed and available (like [this](https://developer.fastly.com/reference/vcl/variables/tcp-info/backend-socket-tcpi-rtt/)) via file descriptors and retrieved via `getsockopt` calls using the file descriptor value. While the Listener filters can access [this value](https://github.com/envoyproxy/envoy/blob/08464ecdc0c93846f3d039d0f0c6fed935f5bdc8/source/extensions/filters/listener/tls_inspector/tls_inspector.cc#L96), our desire is to record this value on a per-request basis.

What do you think about having the socket file descriptor accessible via the HTTP filter callbacks? 

While one possibility would be to store this information in the Listener filter state and access it via the HTTP filter connection’s StreamInfo data, there isn’t a way of invoking a function in the context of the Listener filter to retrieve the TCP-level info from the underlying socket’s file descriptor at the time of the request.

Another possible option is to use HTTP/2 `PING` frames to measure rtt, but this would only apply to clients that speak HTTP/2 and in our use-case, the upstreams clients control the protocol being used.

This use-case was also discussed a bit in https://github.com/envoyproxy/envoy/issues/1464.

### PR DESCRIPTION
**Pull request by davinci26**
Signed-off-by: Sotiris Nanopoulos <sonanopo@microsoft.com>

Commit Message:

Makes an abstraction layer for `TCP_INFO` that is cross-platform and fixes #12305.
Also adds an integration test to make sure that the system calls are correct.

Risk Level: Low
Testing: Tests
Docs Changes: N/A
Release Notes: N/A

### CODE REVIEW

File: include/envoy/common/platform.h
Comment (davinci26) [inline include/envoy/common/platform.h:28]: @wrowe  this is the hack I was talking about offline. If you comment out line 22 and 23 then this static assert hits

**Comment by davinci26** [comment]
cc: @envoyproxy/windows-dev

File: include/envoy/common/platform.h
Comment (nigriMSFT) [inline include/envoy/common/platform.h:28]: Do we know when exactly this was introduced? If it was roughly around the time afunix.h was introduced, do we need a similar comment as the afunix one below?

File: include/envoy/common/platform.h
Comment (nigriMSFT) [inline include/envoy/common/platform.h]: refer to the Windows SIO_TCP_INFO documentation

File: include/envoy/common/platform.h
Comment (davinci26) [reply include/envoy/common/platform.h:28]: It is a bit different than `<afunix.h>` since with `afunix.h` the whole header will be missing while here only one definition in missing. In the later case we can add a static assert.

When we fix the version issue I can add a static assert. So the logic will go as follows:

1. include `#include <mstcpip.h>`
2. If your version of the sdk is old then `SIO_TCP_INFO` will not be defined since the definition is guarded by:
```#if (NTDDI_VERSION >= NTDDI_WIN10_RS2)
#define SIO_TCP_INFO                        _WSAIORW(IOC_VENDOR,39)
#endif // NTDDI_VERSION >= NTDDI_WIN10_RS2
```
3. The `static_assert(false, "SIO_TCP_INFO not defined");` will be hit with a nice error message.

File: include/envoy/common/platform.h
Comment (davinci26) [reply include/envoy/common/platform.h]: Sure I can add the pointer to the Windows documentation but anyone would still need to go though the Windows implementation to verify that we have implemented the mapping from SIO_TCP_INFO -> TCP_INFO.

It might be there is a kernel update for SIO_TCP_INFO and Envoy implementation has not picked up the new version.

File: include/envoy/common/platform.h
Comment (davinci26) [reply include/envoy/common/platform.h:28]: For this one I will leave the comment as pending until we figure out how the headers order should work properly and then I will add it as a static assert.

File: test/common/network/io_socket_handle_impl_test.cc
Comment (nigriMSFT) [inline test/common/network/io_socket_handle_impl_test.cc:100]: Is this true on both platforms?

Comment (davinci26) [reply test/common/network/io_socket_handle_impl_test.cc:100]: I got this based on a comment on slack for linux, haven't personally tried it. On Windows that I tried it did not work with not supported error.

File: test/common/network/io_socket_handle_impl_test.cc
Comment (nigriMSFT) [inline test/common/network/io_socket_handle_impl_test.cc:102]: Is there precedence in the Envoy tests for connecting to remote endpoints? This adds an internet connectivity dependency, which could be flaky.

File: include/envoy/common/platform.h
Comment (nigriMSFT) [reply include/envoy/common/platform.h]: Oh yeah I see. Looking at the Windows OsSysCallsImpl::socketTcpInfo makes more sense and gives a faster, concrete answer

File: test/common/network/io_socket_handle_impl_test.cc
Comment (davinci26) [reply test/common/network/io_socket_handle_impl_test.cc:102]: Fair point and open for further discussion. 

The reasons I went with a remote endpoint:

1. If we only mock this functionality given that this code is optional it becomes a big blind spot. If a regression is introduced in this call it will be extremely hard to detect.
2. I assume public dns servers (either from cloudflare or google) are highly available.
3. It is easier to be optimistic and disable the test if it turns out to reduce developer velocity than to add it afterwards.
4. There is no good alternative to test on localhost

File: include/envoy/common/platform.h
Comment (davinci26) [reply include/envoy/common/platform.h]: Added a link to the SIO_TCP_INFO documentation.

File: test/common/network/io_socket_handle_impl_test.cc
Comment (nigriMSFT) [reply test/common/network/io_socket_handle_impl_test.cc:102]: Is there a way to somehow annotate this test as optional or manually run only? Based on the slack discussion it looked like Yan also had some reservations. I would agree that its nice to have the test in place anytime someone needs to work in the space and verify their changes, but a flaky automated test will also lower productivity and cause frustration. FWIW I tried searching the test code and couldn't really find a concrete example of connecting to remote endpoints

Comment (davinci26) [reply test/common/network/io_socket_handle_impl_test.cc:102]: @envoyproxy/senior-maintainers any thoughts/preference on this?

File: include/envoy/common/platform.h
Comment (nigriMSFT) [inline include/envoy/common/platform.h:23]: Shouldn't NTDDI_VERSION be defined to the minimum system version required (in Envoy case, whatever version afunix.h was introduced)? Looks like WDK_NTDDI_VERSION represents "latest". https://docs.microsoft.com/en-us/windows/win32/winprog/using-the-windows-headers

**Comment by wrowe** [comment]
Nothing should interfere with using the most modern rev numbers, but we will want to delay load the newest API's based on a runtime revision check. I'd rather we weren't compiling several flavors for different OS SDK revisions.

File: include/envoy/common/platform.h
Comment (davinci26) [reply include/envoy/common/platform.h]: Based on the help of @nigriMSFT offline, the previous definition was at `envoy_internal.bzl` which was for an older windows version. I have bumped it to windows rs4 which according to the sources [build to version mapping](https://support.lenovo.com/ca/en/solutions/ht502786) and [af_unix announcement](https://devblogs.microsoft.com/commandline/af_unix-comes-to-windows/) is the minimum version that we need.

File: include/envoy/common/platform.h
Comment (davinci26) [reply include/envoy/common/platform.h]: see comment above. We changed the `NTDDI_VERSION` to the version that afunix was introduced

File: include/envoy/common/platform.h
Comment (nigriMSFT) [inline include/envoy/common/platform.h:311]: nit: `// The corresponding Windows API is documented at:`

Comment (davinci26) [reply include/envoy/common/platform.h]: thanks resolved

**Comment by davinci26** [comment]
@wrowe I expect this to go green, can I get a review from you and @yanavlasov to get this merged

File: source/common/api/win32/os_sys_calls_impl.cc
Comment (sunjayBhatia) [inline source/common/api/win32/os_sys_calls_impl.cc:377]: nit: not big deal, but you could get rid of the else case

File: test/common/network/io_socket_handle_impl_test.cc
Comment (sunjayBhatia) [reply test/common/network/io_socket_handle_impl_test.cc:102]: maybe tag this as a `manual` test?

File: bazel/envoy_internal.bzl
Comment (sunjayBhatia) [inline bazel/envoy_internal.bzl:29]: Should we document this somewhere? That you will need a pretty up to date Windows 10/Server 2019 version?

**Review by sunjayBhatia — COMMENTED** [review]
Added a few comments, overall looks good

File: bazel/envoy_internal.bzl
Comment (davinci26) [reply bazel/envoy_internal.bzl:29]: RS4 was released at Q1 of 2018. Compilation will fail before that anyway because the `af_unix` header does not exist.  Don't we do that already?

Comment (sunjayBhatia) [reply bazel/envoy_internal.bzl:29]: ok sweet, that sounds good to me then if RS4 is old enough

File: include/envoy/api/os_sys_calls.h
Comment (yanavlasov) [inline include/envoy/api/os_sys_calls.h:177]: In the current use case, we just need RTT value. I think at this point it would make more sense to define EnvoyTcpInfo with just `RTT` member. It was suggested in the original PR but I think it makes a lot more sense now. In this case you do not need to pull in tcp_info struct into Envoy's headers.

**Review by yanavlasov — CHANGES_REQUESTED** [review]

File: include/envoy/api/os_sys_calls.h
Comment (davinci26) [reply include/envoy/api/os_sys_calls.h:177]: I am planning to add the header in this header? Is there a better place for small structs?

File: bazel/envoy_internal.bzl
Comment (davinci26) [reply bazel/envoy_internal.bzl:29]: Added also a comment in the readme file.

File: source/common/api/win32/os_sys_calls_impl.cc
Comment (davinci26) [reply source/common/api/win32/os_sys_calls_impl.cc]: fixed in 7c9137b7a3fb86ccae63cbd4018683cc47b7cb4d

File: test/common/network/io_socket_handle_impl_test.cc
Comment (davinci26) [reply test/common/network/io_socket_handle_impl_test.cc]: fixed in 7c9137b7a3fb86ccae63cbd4018683cc47b7cb4d

File: include/envoy/api/os_sys_calls.h
Comment (davinci26) [reply include/envoy/api/os_sys_calls.h]: Fixed in 7c9137b7a3fb86ccae63cbd4018683cc47b7cb4d.

Feel free to open the conversation again if you think that the struct should be moved into a different header

File: bazel/README.md
Comment (davinci26) [inline bazel/README.md:119]: see here @envoyproxy/windows-dev

**Comment by wrowe** [comment]
Can we translate this to a minimum version number of the Windows Platform SDK they would recognize in the SDK, Visual Studio or BuildTools installation kits?

**Comment by davinci26** [comment]
I think, I will just add both. 

Note: These instructions apply to **Windows 10 SDK, version 1803 (10.0.17134.12)**. Earlier versions will not compile because the `afunix.h` header is not available. **The recommended Windows version is equal or later than Windows 10 SDK, version 1903 (10.0.18362.1)**

**Comment by davinci26** [comment]
btw the source of truth for the versions is this link: https://developer.microsoft.com/en-us/windows/downloads/sdk-archive/

**Comment by davinci26** [comment]
@yanavlasov with the new changes (moving the test to separate file with manual tag) the coverage of the systems_calls header has dropped and I dont think it is possible to increase it. Any idea how I can work around it?

File: test/common/network/io_socket_handle_impl_test.cc
Comment (lizan) [reply test/common/network/io_socket_handle_impl_test.cc]: Don't tag it `manual`, that prevent this test from built, just tag this `requires-network`, as documented in https://docs.bazel.build/versions/master/be/common-definitions.html

**Comment by davinci26** [comment]
@yanavlasov CI is all good now, waiting for another round of review

**Review by wrowe — DISMISSED** [review]
This looks good to me, @yanavlasov it appears all of your issues are addressed?

**Review by yanavlasov — APPROVED** [review]

File: source/common/network/io_socket_handle_impl.cc
Comment (yanavlasov) [inline source/common/network/io_socket_handle_impl.cc:530]: The comment for `info.tcpi_rtt` says it is in microseconds.

**Review by yanavlasov — CHANGES_REQUESTED** [review]

File: include/envoy/api/os_sys_calls.h
Comment (yanavlasov) [inline include/envoy/api/os_sys_calls.h:16]: You could potentially make it std::chrono::microseconds type, or add a suffix to indicate units to minimize conversion problems. Strongly typing is probably better.

**Review by yanavlasov — CHANGES_REQUESTED** [review]

File: source/common/network/io_socket_handle_impl.cc
Comment (davinci26) [reply source/common/network/io_socket_handle_impl.cc]: fixed in a4c473a91d2bcc3bdc15ce56f6a7dcc7bbad7556

File: include/envoy/api/os_sys_calls.h
Comment (davinci26) [reply include/envoy/api/os_sys_calls.h]: fixed in a4c473a91d2bcc3bdc15ce56f6a7dcc7bbad7556. I added  `std::chrono::microseconds` in the struct to be consistent with the unix struct. But I maintained the backwards compatibility in the api because I assume that would break the consumers of the function outside of envoy

**Review by yanavlasov — DISMISSED** [review]

File: source/common/network/io_socket_handle_impl.cc
Comment (mattklein123) [inline source/common/network/io_socket_handle_impl.cc:525]: I would make this debug level vs. warn to avoid log spam (or better just drop this entirely and let the caller log/warn, etc. as they desire).

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Small comment, thank you!

/wait

File: source/common/network/io_socket_handle_impl.cc
Comment (davinci26) [reply source/common/network/io_socket_handle_impl.cc]: sure, fixed in 6de9beb3436f68386bcf19f0d1dd34b1bdffa56b

**Review by mattklein123 — APPROVED** [review]
Thanks!

**Pull request merged**
